feature/google/src/bc5SupportDev/java/com/revenuecat/purchases/google/BillingWrapper.kt

```diff
@@ -0,0 +1,728 @@
+//  Purchases
+//
+//  Copyright © 2019 RevenueCat, Inc. All rights reserved.
+//
+
+package com.revenuecat.purchases.google
+
+import android.app.Activity
+import android.content.Context
+import android.os.Handler
+import androidx.annotation.UiThread
+import androidx.annotation.VisibleForTesting
+import androidx.annotation.VisibleForTesting.PRIVATE
+import com.android.billingclient.api.AcknowledgePurchaseParams
+import com.android.billingclient.api.BillingClient
+import com.android.billingclient.api.BillingClient.SkuType
+import com.android.billingclient.api.BillingClientStateListener
+import com.android.billingclient.api.BillingFlowParams
+import com.android.billingclient.api.BillingResult
+import com.android.billingclient.api.ConsumeParams
+import com.android.billingclient.api.Purchase
+import com.android.billingclient.api.PurchaseHistoryRecord
+import com.android.billingclient.api.PurchaseHistoryResponseListener
+import com.android.billingclient.api.PurchasesUpdatedListener
+import com.android.billingclient.api.SkuDetailsParams
+import com.android.billingclient.api.SkuDetailsResponseListener
+import com.revenuecat.purchases.ProductType
+import com.revenuecat.purchases.PurchasesError
+import com.revenuecat.purchases.PurchasesErrorCallback
+import com.revenuecat.purchases.PurchasesErrorCode
+import com.revenuecat.purchases.common.BillingAbstract
+import com.revenuecat.purchases.common.LogIntent
+import com.revenuecat.purchases.common.ReplaceSkuInfo
+import com.revenuecat.purchases.common.StoreProductsCallback
+import com.revenuecat.purchases.common.caching.DeviceCache
+import com.revenuecat.purchases.common.errorLog
+import com.revenuecat.purchases.common.firstSku
+import com.revenuecat.purchases.common.listOfSkus
+import com.revenuecat.purchases.common.log
+import com.revenuecat.purchases.common.sha1
+import com.revenuecat.purchases.common.sha256
+import com.revenuecat.purchases.common.toHumanReadableDescription
+import com.revenuecat.purchases.models.PurchaseState
+import com.revenuecat.purchases.models.StoreProduct
+import com.revenuecat.purchases.models.StoreTransaction
+import com.revenuecat.purchases.models.skuDetails
+import com.revenuecat.purchases.strings.BillingStrings
+import com.revenuecat.purchases.strings.OfferingStrings
+import com.revenuecat.purchases.strings.PurchaseStrings
+import com.revenuecat.purchases.strings.RestoreStrings
+import java.io.PrintWriter
+import java.io.StringWriter
+import java.util.concurrent.ConcurrentLinkedQueue
+import kotlin.math.min
+
+private const val RECONNECT_TIMER_START_MILLISECONDS = 1L * 1000L
+private const val RECONNECT_TIMER_MAX_TIME_MILLISECONDS = 1000L * 60L * 15L // 15 minutes
+
+class BillingWrapper(
```

Comment:
Much of this code seems to be shared between the `latestDependencies` and `bc5SupportDev` versions. Looking at the diff, most changes seem to be around the `BillingType` vs `SkuType` migration currently. I'm sure this will change as we do more changes, but I wonder if we should try to keep as much of the code in this class unified in the `main` package and just split the code as needed.

That said, I'm not fully familiar with all the changes, and it's possible we end up with a completely different class in the end. If so, I think this is ok.

Reply:
There is another, much larger refactor coming for `querySkuDetails` which will replace all `SkuDetails` with `ProductDetails` and will propagate into our `StoreProduct` model/potentially other data models. I was kinda torn whether I should try to unify common things now, or keep them separate and then split them up when I see the full diff. I definitely agree we should reuse as much code as possible, and that some unification needs to happen at some point... 

I'm tempted to say let's wait until all the changes are in. My main argument for that is that if I started touching `latestDependencies` code, I'd want to thoroughly (manually) test that purchases work as expected for each PR. Since we probably need a major round of testing at the end before releasing anyway, I'm inclined to say let's put that off. Though maybe the upcoming automated tests will make this kind of refactoring less scary.. 

Another reason is that what's "common code" right now will definitely change as more refactors come through. So factoring things out now could be throwaway work.

That's my brain dump. Let me know your thoughts, totally open to having my mind changed

Reply:
Yeah makes sense! 👍